connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java

```diff
@@ -357,6 +367,62 @@ public void validateConnectorConfig(Map<String, String> connectorProps, Callback
         });
     }
 
+    /**
+     * Build the {@link RestartPlan} that describes what should and should not be restarted given the restart request
+     * and the current status of the connector and task instances.
+     *
+     * @param request the restart request; may not be null
+     * @return the restart plan, or empty if this worker has no status for the connector named in the request and therefore the
+     *         connector cannot be restarted
+     */
+    public Optional<RestartPlan> buildRestartPlan(RestartRequest request) {
+        String connectorName = request.connectorName();
+        ConnectorStatus connectorStatus = statusBackingStore.get(connectorName);
+        if (connectorStatus == null) {
+            return Optional.empty();
+        }
+
+        // If requested, mark the connector as restarting
+        AbstractStatus.State connectorState;
+        if (request.shouldRestartConnector(connectorStatus)) {
+            connectorState = AbstractStatus.State.RESTARTING;
+        } else {
+            connectorState = connectorStatus.state();
+        }
+        ConnectorStateInfo.ConnectorState connectorInfoState = new ConnectorStateInfo.ConnectorState(
+                connectorState.toString(),
+                connectorStatus.workerId(),
+                connectorStatus.trace()
+        );
+
+        // Collect the task states, If requested, mark the task as restarting
+        List<ConnectorStateInfo.TaskState> taskStates = statusBackingStore.getAll(connectorName)
+                .stream()
+                .map(taskStatus -> {
+                    AbstractStatus.State state = taskStatus.state();
+                    if (request.shouldRestartTask(taskStatus)) {
+                        state = AbstractStatus.State.RESTARTING;
+                    }
```

Comment:
A bit confusing that a second assignment follows if the `if` statement is true. 
I'd also call the variable `taskState` (as opposed to `connectorState` above)

Ternary can be used here as well: 
```suggestion
                    AbstractStatus.State state = request.shouldRestartTask(taskStatus)
                            ? AbstractStatus.State.RESTARTING
                            : taskStatus.state();
```
(as with any suggestion from github, please check it compiles and conforms to the style)